plugins/askrene/layer.c

```diff
@@ -0,0 +1,475 @@
+#include <plugins/askrene/layer.h>
+
+/* A channels which doesn't (necessarily) exist in the gossmap. */
+struct local_channel {
+	/* Canonical order, n1 < n2 */
+	struct node_id n1, n2;
+	struct short_channel_id scid;
+	struct amount_msat capacity;
+
+	struct added_channel_half {
+		/* Other fields only valid if this is true */
+		bool enabled;
+		u16 delay;
+		u32 proportional_fee;
+		struct amount_msat base_fee;
+		struct amount_msat htlc_min, htlc_max;
+	} half[2];
+};
+
+static const struct constraint_key *
+constraint_key(const struct constraint *c)
+{
+	return &c->key;
+}
+
+static size_t hash_constraint_key(const struct constraint_key *key)
+{
+	/* scids cost money to generate, so simple hash works here */
+	return (key->scidd.scid.u64 >> 32) ^ (key->scidd.scid.u64)
+		^ (key->scidd.dir << 1) ^ (key->type);
+}
+
+static inline bool constraint_eq_key(const struct constraint *c,
+				     const struct constraint_key *key)
+{
+	return short_channel_id_dir_eq(&key->scidd, &c->key.scidd) && key->type == c->key.type;
+}
+
+HTABLE_DEFINE_TYPE(struct constraint, constraint_key, hash_constraint_key,
+		   constraint_eq_key, constraint_hash);
+
+static struct short_channel_id
+local_channel_scid(const struct local_channel *lc)
+{
+	return lc->scid;
+}
+
+static size_t hash_scid(const struct short_channel_id scid)
+{
+	/* scids cost money to generate, so simple hash works here */
+	return (scid.u64 >> 32) ^ (scid.u64 >> 16) ^ scid.u64;
+}
+
+static inline bool local_channel_eq_scid(const struct local_channel *lc,
+					 const struct short_channel_id scid)
+{
+	return short_channel_id_eq(scid, lc->scid);
+}
+
+HTABLE_DEFINE_TYPE(struct local_channel, local_channel_scid, hash_scid,
+		   local_channel_eq_scid, local_channel_hash);
+
+struct layer {
+	/* Inside global list of layers */
+	struct list_node list;
+
+	/* Unique identifiers */
+	const char *name;
+
+	/* Completely made up local additions, indexed by scid */
+	struct local_channel_hash *local_channels;
+
+	/* Additional info, indexed by scid+dir */
+	struct constraint_hash *constraints;
+
+	/* Nodes to completely disable (tal_arr) */
+	struct node_id *disabled_nodes;
+};
+
+struct layer *new_temp_layer(const tal_t *ctx, const char *name)
+{
+	struct layer *l = tal(ctx, struct layer);
+
+	l->name = tal_strdup(l, name);
+	l->local_channels = tal(l, struct local_channel_hash);
+	local_channel_hash_init(l->local_channels);
+	l->constraints = tal(l, struct constraint_hash);
+	constraint_hash_init(l->constraints);
+	l->disabled_nodes = tal_arr(l, struct node_id, 0);
+
+	return l;
+}
+
+struct layer *new_layer(struct askrene *askrene, const char *name)
+{
+	struct layer *l = new_temp_layer(askrene, name);
+	list_add(&askrene->layers, &l->list);
+	return l;
+}
+
+/* Swap if necessary to make into BOLT-7 order.  Return direction. */
+static int canonicalize_node_order(const struct node_id **n1,
+				   const struct node_id **n2)
+{
+	const struct node_id *tmp;
+
+	if (*n1 < *n2)
```

Comment:
Comparing pointers' addresses here?
I think you meant `node_id_cmp`.

```suggestion
	if (node_id_cmp(*n1, *n2)<0)
```

Reply:
Woah!  That will teach me to jump between Python and C so much.  What a brain fart, good catch!